### ISSUE TITLE
Proposal: Use a combination of `test` and `remove` JSON Patch operations instead of treating to-many relationships like sets

### ISSUE THREAD

**Issue by fancyremarker** [json-api/json-api#176]
Issue #100 (and the discussion on #99 also) reveal an unresolved question in the spec, around how to properly form JSON Patch operations that remove an item from a to-many relationship. Currently, the spec says:

> While to-many relationships are represented as a JSON array in a GET response, they are updated as if they were a set. [...] To remove an element from a to-many relationship, use a `remove` operation on `links/<name>/<id>`.

This is problematic because JSON Pointer ([RFC 6901](http://tools.ietf.org/html/rfc6901)), the address format for JSON Patch's ([RFC 6902](http://tools.ietf.org/html/rfc6902)) `path`, makes no mention of sets, only arrays. Consider the following example resource, a post with 4 comments:

``` json
{
  "posts": [{
    "id": "1",
    "links": {
      "comments": [ "1", "3", "5", "7" ]
    }
  }]
}
```

The current JSON API spec says that if we want to remove the comment with ID 3, we'd make the following `PATCH` request:

```
PATCH /photos/1

[
  { "op": "remove", "path": "/posts/0/links/comments/3" }
]
```

This is a valid JSON Patch document, but its standard semantics are different from what the JSON API spec would have it mean. A naive client built on a compliant JSON  Patch library could very easily make the above request with the intention of removing the element with index 3 in the `comments` array (i.e., comment 7), but instead, our JSON API-compliant server would remove the wrong comment, comment 3.

I propose that in order to remove an element from a to-many relationship, a client must use a `remove` operation on `links/<name>/<arrayIndex>`, along with a `test` that `links/<name>/<arrayIndex>` has the value we expect. With our example from above, this would look like:

```
PATCH /photos/1

[
  { "op": "test", "path": "/posts/0/links/comments/1", "value": "3" }
  { "op": "remove", "path": "/posts/0/links/comments/1" }
]
```

This is guaranteed to always be safe, and also complies with both the JSON Patch and JSON Pointer specs. For more info on the `test` operation see [RFC 6902, §4.6](http://tools.ietf.org/html/rfc6902#section-4.6).

**Comment by dgeb** [comment]
I think we should clear something up in the spec:

_Arrays of documents are functionally equivalent to sets that are keyed by document ids._

In RFC 6901 / 6902 terms, a set is simply an object with "members" and "member values".

So, even though responses are structured as:

```
{
  "posts": [{
    "id": "1",
    "links": {
      "comments": [ "1", "3", "5", "7" ]
    }
  }]
}
```

This is functionally equivalent to an RFC6902-compliant document structured like this:

``` javascript
{
  "posts": {
    "1": {
      "id": "1",
      "links": {
        "comments": {
          "1": "1",
          "3": "3",
          "5": "5",
          "7": "7"
        }
      }
    }
  }
}
```

For this structure, removing the comment with id 5 from the post with id 1 can be done with an RFC6902-compliant operation:

```
PATCH /posts/1

[
  { "op": "remove", "path": "/posts/1/links/comments/5" }
]
```

I would argue that since we're making PATCH operations against equivalent instead of actual structures, it might even be preferable to drop `links` from the path (and functionally equivalent representation):

```
PATCH /posts/1

[
  { "op": "remove", "path": "/posts/1/comments/5" }
]
```

This is a common point of confusion, and I totally understand where you're coming from by constructing your operations and paths based on patching arrays and not objects. However, accessing arrays of distributed data by position just doesn't work well in practice. Your `test` operation would fail if another object had been inserted in the remote set before this one causing its position to shift. You'd then have to request a new remote array and try again. I think it's a lot more forgiving and efficient to treat arrays as sets and skip the `test` step.

So, why not restructure the JSON response body (as suggested in #154) to match the functionally equivalent document structure? I think mostly for the sake of brevity and ease of navigation.

Bottom line: I think we should stick with the structure outlined currently (with the exception of removing `links` from JSON pointer paths) and spend some time clarifying how the actual JSON response body maps to paths accessible with JSON pointers.

**Comment by mattupstate** [comment]
Why do you then need `/posts/1` in the path?

**Comment by dgeb** [comment]
@mattupstate Good catch. Either of the following would probably be better:

```
PATCH /posts/1

[
  { "op": "remove", "path": "/comments/5" }
]
```

or 

```
PATCH /

[
  { "op": "remove", "path": "/posts/1/comments/5" }
]
```

It comes down to whether to allow absolute or relative paths, or both.

**Comment by mattupstate** [comment]
Another thing...because I'm ignorant (just getting into this stuff). I have to ask...what if my ID's are opaque? Does this still apply?

```
PATCH /posts/xyz123

[
  { "op": "remove", "path": "/posts/1/comments/5" }
]
```

I feel like there's a slight cognitive disconnect there. Maybe its just me.

**Comment by guillec** [comment]
@mattupstate going to try to answer this one to make sure I am understanding what is being proposed :)

Your patch document should then actually look like 

```
[
  { "op": "remove", "path": "/posts/xyz123/comments/5" }
]

or

[
  { "op": "remove", "path": "/comments/5" }
]
```

because it would have to match

```
{
  "posts": {
    "xyz123": {
      "id": "1",
      "links": {
        "comments": {
          "1": "1",
          "3": "3",
          "5": "5",
          "7": "7"
        }
      }
    }
  }
}
```

**Comment by mattupstate** [comment]
@guillec That makes sense, then. Although is there a reason you kept the attribute `"id": "1"` in there?

**Comment by guillec** [comment]
Woops copy, paste and update error! That should be xyz123.

**Comment by dgeb** [comment]
@guillec thanks for clarifying that :+1: 

I should also expand upon what I said above:

> So, why not restructure the JSON response body (as suggested in #154) to match the functionally equivalent document structure? I think mostly for the sake of brevity and ease of navigation.

As @mattupstate points out in https://github.com/json-api/json-api/issues/154#issuecomment-33822869, we also need to keep collections structured as arrays to support edge cases where IDs are not included in documents. (Note: documents without IDs can only be retrieved and should never be pushed back to the server).

**Comment by MajorBreakfast** [comment]
As @dgeb and @mattupstate pointed out there is a problem. What if

`PATCH /`:

``` JSON
[{ "op": "remove", "path": "/posts/1/comments/5" }]
```

`PATCH /posts/1`:

``` JSON
[{ "op": "remove", "path": "/comments/5" }]
```

... did the same thing? (I also left out the "links/" part)

**Comment by dgeb** [comment]
@MajorBreakfast I agree those requests should be equivalent and could optionally both be supported by an API. I actually just [commented](https://github.com/json-api/json-api/issues/193#issuecomment-34757572) to that effect over in #193.

**Comment by steveklabnik** [comment]
Adding to 1.0.

**Comment by steveklabnik** [comment]
PATCH today is in an extension, and so I'm de-milestoning for 1.0.

**Comment by tkellen** [comment]
This was resolved in #237.
